subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/ivy/RealisedIvyModuleResolveMetadata.java

```diff
@@ -234,7 +236,15 @@ public MutableIvyModuleResolveMetadata asMutable() {
 
     @Override
     public RealisedIvyModuleResolveMetadata withSources(ModuleSources sources) {
-        return new RealisedIvyModuleResolveMetadata(this, sources);
+        return new RealisedIvyModuleResolveMetadata(this, sources, getVariantDerivationStrategy());
+    }
+
+    @Override
+    public ModuleComponentResolveMetadata withDerivationStrategy(VariantDerivationStrategy derivationStrategy) {
+        if (getVariantDerivationStrategy() == derivationStrategy) {
```

Comment:
I would use `.equals` here. We know today that the instance equality is enough, but if this changes down the line, it might be surprising.